src/goto-checker/multi_path_symex_checker.cpp

```diff
@@ -0,0 +1,238 @@
+/*******************************************************************\
+
+Module: Goto Checker using Bounded Model Checking
+
+Author: Daniel Kroening, Peter Schrammel
+
+\*******************************************************************/
+
+/// \file
+/// Goto Checker using Bounded Model Checking
+
+#include "multi_path_symex_checker.h"
+
+#include <chrono>
+
+#include "bmc_util.h"
+#include "counterexample_beautification.h"
+
+multi_path_symex_checkert::multi_path_symex_checkert(
+  const optionst &options,
+  ui_message_handlert &ui_message_handler,
+  abstract_goto_modelt &goto_model)
+  : multi_path_symex_only_checkert(options, ui_message_handler, goto_model),
+    symex_has_run(false)
+{
+  solver_factoryt solvers(
+    options,
+    ns,
+    ui_message_handler,
+    ui_message_handler.get_ui() == ui_message_handlert::uit::XML_UI);
+  solver = solvers.get_solver();
+  solver->prop_conv().set_message_handler(ui_message_handler);
+}
+
+incremental_goto_checkert::resultt multi_path_symex_checkert::
+operator()(propertiest &properties)
+{
+  resultt result;
+
+  if(!symex_has_run)
+  {
+    perform_symex();
+    output_coverage_report();
+    result.updated_properties =
+      update_properties_status_from_symex_target_equation(properties, equation);
+  }
+
+  // have we got anything to check? otherwise we return DONE
+  if(!has_properties_to_check(properties))
+    return result;
+
+  prop_convt &prop_conv = solver->prop_conv();
+
+  auto solver_start = std::chrono::steady_clock::now();
+
+  if(!symex_has_run)
+  {
+    log.status() << "Passing problem to " << prop_conv.decision_procedure_text()
+                 << messaget::eom;
+
+    convert_symex_target_equation(equation, prop_conv, ui_message_handler);
+    update_properties_goals_from_symex_target_equation(properties);
```

Comment:
Would it be possible to document what happens when this `operator()` is executed repeatedly, and whether there are any assertions that must be met on repeated execution?

The reason I am asking is that this line seems to do something to the equation depending on that argument is given as `properties`, but the line will not be executed on the second call to `operator()`, which can pass a different argument than the first call. Is this intended?

Reply:
To @tautschnig's delight, I have now added documentation to this function to clarify this behavior.

Reply:
I'm sorry, but I still don't get it. What behaviour do you expect if somebody calls `operator()` twice in a row, each time with a different argument. Say I have properties `p1` and `p2`, each individually stored in `propertiest` object `pp1` and pp2` respectively, i.e. the containers are disjoint. Then I call
```
multi_path_symex_checkert(pp1); //here I get what is happening, p1 will be set up correctly, but p2 will be untouched as far as I can tell
multi_path_symex_checkert(pp2); //here I think all setup for p2 will be skipped, and I don't see why that is ok. 
```